packages/contracts-bedrock/src/L2/Callbacks.sol

```diff
@@ -0,0 +1,161 @@
+            revert("Greeter_InvalidGreeting");
+        }
+
+        greeting = _greeting;
+        // emit GreetingSet(_greeting);
+    }
+
+    /**
+     * @notice Initiates a remote greeting call handled by the callback
+     * @return contextNonce_
+     */
+    function remoteGreet(CallbackContext calldata _callbackContext) external returns (uint224 contextNonce_) {
+        contextNonce_ = async(
+            2, // TODO: receive param
+            abi.encodeWithSelector(this.greeting.selector),
+            this.remoteGreetCallback.selector,
+            _callbackContext
+        );
+    }
+
+    function remoteGreetCallback(
+        uint224 _contextNonce,
+        string memory _remoteGreeting
+    )
+        external
+        returns (string memory _greeting, uint256 _balance)
+    {
+        _greeting = _remoteGreeting;
+        _balance = address(msg.sender).balance;
+
+        // obtain the return context
+        CallbackContext memory _callbackContext = callbackContexts[_contextNonce];
+        // delete the return context
+        delete callbackContexts[_contextNonce];
+
+        // call the callback (not sure if we should check for success...)
+        _callbackContext.to.call(abi.encodeWithSelector(_callbackContext.selector, _greeting, _balance));
+    }
+
+    // builds the cdm call with the callback entrypoint
+    function async(
+        uint256 _chainid,
+        bytes memory _data,
+        bytes4 _callbackSelector,
+        CallbackContext calldata _callbackContext
+    )
+        internal
+        returns (uint224 contextNonce_)
+    {
+        // increment the nonce
+        contextNonce_ = ++returnContextNonce;
+        // store the return context
+        callbackContexts[contextNonce_] = _callbackContext;
+
+        IL2ToL2CrossDomainMessenger(L2_TO_L2_CROSS_DOMAIN_MESSENGER).sendMessage(
+            _chainid,
+            _remoteGreeter,
+            // data + callback selector
+            abi.encodePacked(_data, _callbackSelector, contextNonce_),
+            //abi.encodePacked(_data, _callbackSelector, contextNonce_, 32), // 32 bytes can be dynamic [IMPROVEMENT]
+            CALLBACK_ENTRYPOINT
+        );
+    }
+
+    // Setter for the remote greeter WIP without auth for now.
+    function setRemoteGreeter(address __remoteGreeter) public {
+        _remoteGreeter = __remoteGreeter;
+    }
+}
+
+contract CallbackEntrypoint {
+    address public constant L2_TO_L2_CROSS_DOMAIN_MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
+
+    function relayMessage(
+        Identifier calldata _id,
+        bytes calldata _sentMessage
+    )
+        external
+        payable
+        returns (bytes memory returnData_)
+    {
+        // Calls the CDM contract and get return value of the function call
+        returnData_ = IL2ToL2CrossDomainMessenger(L2_TO_L2_CROSS_DOMAIN_MESSENGER).relayMessage(_id, _sentMessage);
+
+        (address sender,,) = abi.decode(_sentMessage[128:], (address, bytes, address));
```

Comment:
why 128? could we add a comment on what this line does?

Reply:
Can we just cut the 20bytes of the address?
128:148
Or does that not work?

Reply:
Yeah, changed to only decode the sender address, has to be 128:160, tho.